spinn_utilities/ranged/ranged_list.py

```diff
@@ -0,0 +1,513 @@
+
+        :return: yields each element one by one
+        """
+        slice_start, slice_stop = self._check_slice_in_range(
+            slice_start, slice_stop)
+
+        # If range based, go through the ranges that intersect the slice
+        if self._ranged_based:
+
+            # Find the first range within the slice
+            ranges = self.iter_ranges()
+            (start, stop, value) = ranges.next()
+            while stop < slice_start:
+                (start, stop, value) = ranges.next()
+
+            # Continue until outside of the slice
+            while start < slice_stop:
+                first = max(start, slice_start)
+                end_point = min(stop, slice_stop)
+                for _ in xrange(end_point - first):
+                    yield value
+                (start, stop, value) = ranges.next()
+
+        # If non-range-based, just go through the values
+        else:
+            for value in self._ranges[slice_start: slice_stop]:
+                yield value
+
+    def iter_ranges(self):
+        """
+        Fast NOT update safe iterator of the ranges
+
+        :return: yields each range one by one
+        """
+        # If range based just yield the ranges
+        if self._ranged_based:
+            for r in self._ranges:
+                yield r
+
+        # If non-range based, build the ranges
+        else:
+            previous_value = self._ranges[0]
+            previous_start = 0
+            for start, value in enumerate(self._ranges):
+                if value != previous_value:
+                    yield (previous_start, start, previous_value)
+                    previous_start = start
+                    previous_value = value
+            yield (previous_start, start + 1, value)
+
+    def iter_ranges_by_slice(self, slice_start, slice_stop):
+        """
+         Fast NOT update safe iterator of the ranges covered by this slice
+
+         Note: The start and stop of the range will be reduced to just the
+         ids inside the slice
+
+         :return: yields each range one by one
+         """
+        slice_start, slice_stop = self._check_slice_in_range(
+            slice_start, slice_stop)
+
+        # If range-based, go through ranges that intersect the slice
+        if self._ranged_based:
+            for (start, stop, value) in self._ranges:
+                if slice_start < stop:
+
+                    # The range is updated so that the start and stop values
+                    # are within the slice requested
+                    yield (max(start, slice_start), min(stop, slice_stop),
+                           value)
+                    if slice_stop <= stop:
+                        break
+
+        # If non-range based, just go through the values
+        else:
+            for index, value in \
+                    enumerate(self._ranges[slice_start: slice_stop]):
+                yield (slice_start + index, slice_start + index + 1, value)
```

Comment:
In the previous iter_ranges function, a merge of adjacent values is done.   Here, adjacent values result in separate ranges being returned.  This should be consistent.

Reply:
yes to do!

Reply:
Is this to be marked as TODO, or are you intending doing it before the pull request is accepted?  I was worried that this might have a knock-on effect on the get_value calls, since the default implementation checks that there is a single range, however I think the default ends up being overridden in the implementation that would have called this function anyway, so I am happy for this to be left as a TODO.

Reply:
Done!  

It is not critical as ideally this method would never be called when self.ranged_based() is False
But it is better to be consistent.